baremetal/metal3machine_manager.go

```diff
@@ -1741,3 +1791,83 @@ func (m *MachineManager) getNodesWithLabel(ctx context.Context, nodeLabel string
 	}
 	return nodes, nodesCount, err
 }
+
+// getNodesWithoutLabel gets kubernetes nodes based on their Spec.providerID field.
+func (m *MachineManager) getNodesWithoutLabel(ctx context.Context, providerIDLegacy, providerIDNew string, providerIDonM3M *string, clientFactory ClientGetter) (int, error) {
+	corev1Remote, err := clientFactory(ctx, m.client, m.Cluster)
+	matchingNodesCount := 0
+	if err != nil {
+		return matchingNodesCount, errors.Wrap(err, "Error creating a remote client")
+	}
+
+	filter := metav1.ListOptions{}
+	nodes, err := corev1Remote.Nodes().List(ctx, filter)
+	if err != nil {
+		m.Log.Info(fmt.Sprintf("error while retrieving nodes: %v", err))
+		return matchingNodesCount, err
+	}
+	// kubernetes nodes with their providerID and name fields set
+	validNodes := make(map[string][]string)
+	matchingNodeProviderID := ""
+	for _, node := range nodes.Items {
+		providerIDOnNode := node.Spec.ProviderID
+		if node.Spec.ProviderID == "" {
+			m.Log.Info("no providerID value found on node: ", node.GetName())
+		} else if providerIDOnNode == providerIDNew {
+			matchingNodeProviderID = providerIDNew
+		} else if providerIDOnNode == providerIDLegacy {
+			matchingNodeProviderID = providerIDLegacy
+		} else {
+			m.Log.Info("The node: ", node.GetName(), " does not match expected providerID: ", providerIDOnNode, "Considering other nodes ...")
+		}
+		if providerIDOnNode != "" && node.GetName() != "" {
+			validNodes[providerIDOnNode] = append(validNodes[providerIDOnNode], node.GetName())
+		}
+	}
+	err = m.duplicateProviderIDsExist(validNodes, providerIDLegacy, providerIDNew)
+	if err != nil {
+		// There are, at least, two nodes. Details are in err
+		return 2, err
+	}
+	if matchingNodeProviderID != "" {
+		*providerIDonM3M = matchingNodeProviderID
+		return 1, nil
+	}
+	return matchingNodesCount, nil
```

Comment:
Is this matchingNodesCount variable always 0? Because there is no assignment and modification other than its initialization?

Reply:
Yes, it is always zero.
There are multiple locations that should return 0. It is just paramerization.

Reply:
I think if it is always zero, we should not use a variable, but directly return 0. Variable means for something that can be changed.

Reply:
I agree with @namnx228 , just returning 0 would be less confusing. If we still want to have a descriptive name let's make it a constant and name it something like `zeroMatchingNodes`.